change access to clic registers from memory mapped to indirect CSR

dansmathers opened the pull request:
for issue #349, add option for using indirect CSR access method to control clic registers.  only updated text for smclic.  will update for text other priv modes later. Adding indirect CSR access could close #221 issue as well as it would provide a method to quickly enable/disable sets of interrupt enables.

dansmathers committed:
adding indirect CSR access option to clic registers

for issue #349, add option for using indirect CSR access method to control clic registers.  only updated text for smclic.  will update for text other priv modes later.
Adding indirect CSR access could close #221 issue as well as it would provide a method to quickly enable/disable sets of interrupt enables.

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

christian-herber-nxp commented on clic.adoc:
```diff
@@ -387,9 +387,45 @@ interrupt input.
+  ...
+  0x91F     clicinttrig[31}
+  0x920     reserved for smclicconfig extension
+----
```
It seems that registers are only ever mapped to mireg, and mireg2-6 are unused. For configuration, that would cause quite some overhead. My suggestion would be to extend the table with five additional columns for mireg2-6, and to map registers there.

kasanovic replied:
We need to keep the indexing straightforward for software.  The only resource being consumed is miselect address space, and there is not that much being used, since there are 32 interrupts at each index.

christian-herber-nxp commented on clic.adoc:
```diff
@@ -387,9 +387,45 @@ interrupt input.
+  0x900     clicinttrig[0]
+  0x901     clicinttrig[1]
+  ...
+  0x91F     clicinttrig[31}
```
ends in a curly brace

christian-herber-nxp commented on clic.adoc:
```diff
@@ -387,9 +387,45 @@ interrupt input.
+
+=== CLIC Register Access via Indirect CSR Access 
+
+The Indirect CSR Access allocated range for CLIC is TBD.  These address provided below are relative to the base of that allocated range.
```
Spurious space after TBD.
I think it is clearer to refer to select values rather than addresses. Select values are a kind of address, but its a more precise term in this context.

dansmathers committed:
update indir organization to align with mem-mapped organization

also changed indir/memmap from being parameters to being clic extensions

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

christian-herber-nxp reviewed (approved):
looks good to me

dansmathers committed:
rearranged indirect packing of clicintregs, removed mem-map

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

kasanovic commented on clic.adoc:
```diff
@@ -377,94 +377,144 @@ The existing timer (`mtip`/`stip`), software
+ offset    |     bits      |  register state             |                    |
 
-NOTE: For reserved memory regions, specific trap behavior is not specified. Depending on system bus architecture, the system can ignore the access (e.g., read zero/write ignored) or cause a bus error (usually imprecise interrupt), or some other platform-specific behavior. The "reserved" annotation here implies that future standards might place additional standard registers in that space, and so u ...
+  0x0      |     31:0     |   R or RW   clicintip[31:0]  |  clic interrupt pending for interrupts 31-0 |
```
Should just mark as RW, as we can ignore writes if the interrupt is not edge-triggered.  This way we don't have to try to deal with exceptions on writes to read-only bits in a register.

kasanovic commented on clic.adoc:
```diff
@@ -525,11 +575,9 @@ or 255.
-possible future expansion of the `clicintip` field.
+(`clicintip[__i__]`). 
+Software should assume `clicintip[__i__]=0` means no interrupt pending, and
+`clicintip[__i__]!=0` indicates an interrupt is pending.
```
Now IP is only a single bit, so can reword description.

kasanovic commented on clic.adoc:
```diff
@@ -562,8 +610,8 @@ level-sensitive mode.
-and occupies one byte in the memory map for ease of access. This control bit is
-read-write to enable/disable the corresponding interrupt.  The enable bit is located in bit 0 of the byte. Software should assume clicintie[i]=0 means no interrupt enabled, and clicintie[i]!=0 indicates an interrupt is enabled to accommodate possible future expansion of the clicintie field.
+This control bit is read-write to enable/disable the corresponding interrupt.  
+Software should assume clicintie[i]=0 means no interrupt enabled, and clicintie[i]!=0 indicates an interrupt is enabled.
```
Same comment on IP now being one bit.

dansmathers committed:
updated clicintctrl/attr/ip/ie indirect organization

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

Silabs-ArjanB commented:
@dansmathers The PR title and some of the text in this PR seems to suggest that an _option_ is being added. Reading through the PR it however seems that this is not the case because:

- The original _memory mapped option_ seems completely removed.
- No parameter for this option seems to have been introduced.

There is a statement that _"A CSR accessible via one method may or may not be accessible via the other method such as memory-mapped access"_. This _"other method"_ seems however not specified. 

I would really prefer if the memory mapped options stays explicitly specified.

The indirect CSR access option also opens up the discussion again that was closed here: https://github.com/riscv/riscv-fast-interrupt/issues/225. For indirect CSR accesses will there now again be requirements that writes to clicintie or clicintip are evaluated immediately? I think the answer to this question should be made explicit. If the answer is 'yes' then that introduces a weird incompatibility between the two access methods.

dansmathers commented:
Good question on should both indirect and memory mapped access be in this pull.  I've updated the pull title to reflect the current pull status. originally the pull just added indirect CSR access but you are correct that the current pull now replaces memory mapped access with indirect CSR access.   Hopefully we can have a good discussion on adding the memory mapped method back into the spec and how it should be, e.g. if it would be an extension or parameter or something else.

in the 20231205  fast interrupt TG minutes there was this discussion: 
> Issue #349 – using indirect CSRs for clicintctrl/attr/ip/ie. Hesitate to implement both indirect and CSRs and memory mapped.  choose one or the other. 
AIA – one register to implement.  try to be compatible with AIA?  in terms of msi – keep that consistent.  restrict memory map for software interrupts.  just because have msis, don’t open up the whole memory map to everything. 
Don’t' have to have indirect csrs.  understanding indirect csr as base.  memory map just opened up enough to do software interrupts. 


for your question about issue #225:
I still see this statement under clicintip section:
> The conditions for an interrupt trap to occur must be evaluated in a bounded amount of time from when an interrupt becomes, or ceases to be, pending in clicintip, but unlike the MIP/MIE CSRs, there is no requirement that clicintie or clicintip are evaluated immediately following an explicit store to clicintip or clicintie.

in the 20230926 fast interrupt TG minutes, there was this discussion:
> #349 - Fixed csr so pipeline can figure out conflicts.  indirect csrs, don’t know select register.  value may not be ready yet.  so things in ind csrs can’t cause pipeline events.  so means ind csrs aren’t critical timing.  so flexible but more constrained on what they can do.  just like AIA, it is a good candidate.  so ind csrs is like a local memory map

So for this question:
>  For indirect CSR accesses will there now again be requirements that writes to clicintie or clicintip are evaluated immediately?

I would say the answer is no and the answer is explicit in the spec/pull already.  Do you agree?

Silabs-ArjanB commented:
> for your question about issue https://github.com/riscv/riscv-fast-interrupt/issues/225:
> I still see this statement under clicintip section:
> 
> The conditions for an interrupt trap to occur must be evaluated in a bounded amount of time from when an interrupt becomes, or ceases to be, pending in clicintip, but unlike the MIP/MIE CSRs, there is no requirement that clicintie or clicintip are evaluated immediately following an explicit store to clicintip or clicintie.

This section was added because clicintip and clicintie are/were memory mapped. If you change these to be CSR accessible then I think this should be removed and the reverse should be mandated (i.e. they must be evaluated **immediately** following the execution of an xRET instruction or an explicit write to these CSRs).

I think furthermore that mandating indirect CSR access just increases the interrupt latency without any real benefit over using a memory mapped strategy. The reason I think interrupt latency is negatively affected is that the mireg CSR adds state that now needs to be saved and restored (which conflicts with the goal of having an extension that improves interrupt latency).

My proposal would be to mark indirect CSR access as a possible post-v1.0 addition as opposed to a pre-v1.0 replacement of a memory mapped strategy.

christian-herber-nxp commented:
Interesting point regarding the additional state. In most cases, however, I would argue that ISRs would not need to access indirect CSRs, and then no state needs to be saved.

Given the fact that indirect CSR accesses are a possibility now, I think they are the best choice for CLIC, and we want to have the best possible version of CLIC.

Of course I understand that it might be desired to maintain the memory mapped option in the interest of existing implementations, but from software point of view that would essentially lead to two kinds of CLICs (fueling any concerns regarding fragmentation).

I am travelling on Monday and will not be able to join, so please consider this as my input to the discussion.

dansmathers committed:
Add notes about non-atomic indirect CSR access

As commented in #364, indirect CSR access are not atomic and so mireg CSR state needs to be saved and restored if the indirect CSRs are accessed while same privilege mstatus.xie is enabled.   The note was added that this is expected to be atypical but important for software implementers to understand.

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers committed:
Merge branch 'master' into dansmathers-patch-28

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers committed:
added note about indirect CSR context/save restore to ISR example

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers committed:
update pull with updated clicinttrig description

Pull had conflicts after clicinttrig update.  Pull is updated with those changes

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers committed:
change trigger temp to memory-mapped to resolve merge conflict

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers committed:
tmp put MCLICBASE back in to resolve github merge.

Signed-off-by: Dan Smathers <dan.smathers@seagate.com>

dansmathers commented:
I tried to copy comments about moving CLIC memory mapped registers to indirect CSRs to issue #349 for easier issue tracking.  Can we continue the discussion there?
